### PR TITLE
Segmentation

### PR DESCRIPTION
**Pull request by arthurdouillard**
- [x] test for saved indexes
- [x] faster computing of indexes
- [x] logging for computation of indexes
- [x] test with real data?
- [x] in-code doc
- [X] web doc
- [x] make plotting for segmentation
- [x] add citation where it belongs (michielli, cermelli [especially for the code], douillard)
- [x] simplify if in task set
- [x] test for advanced indexing

### CODE REVIEW

File: continuum/tasks/task_set.py [hunk 1]
```diff
@@ -107,7 +110,7 @@ def get_sample(self, index: int) -> np.ndarray:
         """
         x = self._x[index]
 
-        if self.data_type == "image_path":
+        if self.data_type in ("image_path", "segmentation"):
```
Comment (TLESORT) [inline continuum/tasks/task_set.py:113]: pourquoi segmentation et image_path sont deux object de type different ici?

File: continuum/tasks/task_set.py [hunk 2]
```diff
@@ -118,19 +121,25 @@ def get_sample(self, index: int) -> np.ndarray:
...
         """Method used by PyTorch's DataLoaders to query a sample and its target."""
-        sample = self.get_sample(index)
+        x = self.get_sample(index)
         y = self._y[index]
         t = self._t[index]
 
-        if self.data_type != 'text':
-            if self.trsf is not None:
-                sample = self.trsf(sample)
+        if self.data_type == "segmentation":
```
Comment (TLESORT) [inline continuum/tasks/task_set.py:128]: ya bcp de if avec l'ajout de la segmentation, je pense que cree un task_set specialis'e pour la segmentation rendrait les choses plus claire / plus propre

File: continuum/scenarios/base.py [hunk 3]
```diff
@@ -111,7 +123,7 @@ def _select_data_by_task(self, task_index: Union[int, slice]):
             # like PermutedMNIST or RotatedMNIST.
             selected_y = self.cl_dataset.class_remapping(selected_y)
 
-        return selected_x, selected_y, selected_t
+        return selected_x, selected_y, selected_t, task_index
```
Comment (TLESORT) [inline continuum/scenarios/base.py:126]: je comprends pas ce task_index en plus du selected_t, quelque commentaires en plus serait bienvenu :)

File: continuum/scenarios/segmentation.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    :param cl_dataset: A continual dataset.
+    :param nb_tasks: The scenario number of tasks.
+    :param increment: Either number of classes per task (e.g. increment=2),
+                    or a list specifying for every task the amount of new classes
+                     (e.g. increment=[5,1,1,1,1]).
+    :param initial_increment: A different task size applied only for the first task.
+                              Desactivated if `increment` is a list.
+    :param transformations: A list of transformations applied to all tasks.
+    :param class_order: An optional custom class order, used for NC.
+                        e.g. [0,1,2,3,4,5,6,7,8,9] or [5,2,4,1,8,6,7,9,0,3]
```
Comment (TLESORT) [inline continuum/scenarios/segmentation.py:31]: ce commentaire n'est pas 'a jour

File: continuum/scenarios/segmentation.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+from PIL import Image
+import torchvision
+
+from continuum.datasets import _ContinuumDataset
+from continuum.scenarios import ClassIncremental
+from continuum.tasks import TaskSet
+from continuum.download import ProgressBar
+
+
+class SegmentationClassIncremental(ClassIncremental):
```
Comment (TLESORT) [inline continuum/scenarios/segmentation.py:17]: C'est pas possible d'utiliser la class ClassIncremental pour la segmentation pour garder ca transparent pour l'utilisateur ?

File: continuum/tasks/task_set.py (hunk 1 above)
Comment (arthurdouillard) [reply continuum/tasks/task_set.py:113]: While for both `x` are actually image paths, for the `segmentation` type the handling is different in the task set and we rely on the data type for this.

Plus note that `y` is an array of integer when `data_type=image_path`, but an array of paths when `data_type=segmentation`.

File: continuum/tasks/task_set.py (hunk 2 above)
Comment (arthurdouillard) [reply continuum/tasks/task_set.py:128]: I'm not sure abouuy a new task set, it's only a few lines added. But I'll make an effort to simply the number of if, you're right

File: continuum/scenarios/segmentation.py (hunk 4 above)
Comment (arthurdouillard) [reply continuum/scenarios/segmentation.py:31]: Yes, doc is not yet updated, see my todo list in the description of the PR

File: continuum/scenarios/segmentation.py (hunk 5 above)
Comment (arthurdouillard) [reply continuum/scenarios/segmentation.py:17]: That could be a good idea, but we should be able to do it cleanly. I'm going to think about it

File: continuum/scenarios/base.py (hunk 3 above)
Comment (arthurdouillard) [reply continuum/scenarios/base.py:126]: I need it to discernate between multiple tasks. In Seg, a same image can have several tasks.

Yes I'll do doc/commentary :)

File: continuum/tasks/task_set.py (hunk 2 above)
Comment (arthurdouillard) [reply continuum/tasks/task_set.py]: I've simplify the task set handling of the different data types.

**Comment by arthurdouillard** [comment]
I've added the plotting for segmentation.

Right now it relies on the colormapping of Pascal-VOC2012. This PR is already quite big, so different colormapping will be provided in another PR.

File: continuum/scenarios/segmentation.py (hunk 5 above)
Comment (arthurdouillard) [reply continuum/scenarios/segmentation.py:17]: As said in private discussion, segmentation brings too many arguments. It'd be more confusing to merge those two classes together.

File: continuum/tasks/task_set.py (hunk 1 above)
Comment (TLESORT) [reply continuum/tasks/task_set.py:129]: Ok je vois, je reviens a mon idee de faire une class task_set differente pour gerer les donnee de segmentation ca eviterai d'avoir un cree un data_type comme ca un peu arbitraire et pas evident a comprendre de l 'exterieur.

File: continuum/datasets/segmentation.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
+import os
+from typing import Iterable, Set, Tuple, Union
+
+import numpy as np
+
+from continuum import download
+from continuum.datasets.base import _ContinuumDataset
+from continuum.transforms.segmentation import ToTensor
+
+class _SemanticSegmentationDataset(_ContinuumDataset):
```
Comment (TLESORT) [inline continuum/datasets/segmentation.py:10]: je mettrais plutot cette classes dans le fichier base ou dans un fichier base_segmentation.py et renommerai ce fichier pascalvoc ou seg_pascalvoc

File: docs/_tutorials/scenarios/scenarios.rst [hunk 7]
```diff
@@ -310,6 +310,60 @@ supports this setting.
...
+for segmentation is very different from previous scenarios.
+
+Semantic segmentation aims at classifying all pixels in an image, therefore
+multiple classes can co-exist in the same image. This distinction leads to three kinds of scenarios:
+
+- **Sequential**: where for a given task T, with current classes C, the model sees all
+  images that contain at least one pixel labeled as a current classes C. If the image contains
+  future classes, yet unseen, then it is discarded. In the sequential setting, all pixels
+  are labeled, either with a old or current class label, background label (0), or
+  unknown label (255).
```
Comment (TLESORT) [inline docs/_tutorials/scenarios/scenarios.rst:327]: c'est pas un peu cheaté la label "je sais pas encore ce que c'est mais dans le futur on me dira" ?

File: docs/_tutorials/scenarios/scenarios.rst [hunk 8]
```diff
@@ -310,6 +310,60 @@ supports this setting.
...
+- **Disjoint**: It's the same scenario as Sequential, but on one point. An image's pixel is
+  only labeled for the current classes. Therefore, during training, if an old class is
+  present in the image, its labels would be 0 (aka background). However, during the test
+  phase, all labels (current + old) are present.
+- **Overlap**: It's the same scenario as Disjoint except that the model can also
+  see images containing a future class, as long as a current class is present.
+
```
Comment (TLESORT) [inline docs/_tutorials/scenarios/scenarios.rst:334]: je suis pas sur de tout comprendre dans ces 3 scenarios, ca me parait pas tres claire

File: docs/_tutorials/scenarios/scenarios.rst (hunk 7 above)
Comment (arthurdouillard) [reply docs/_tutorials/scenarios/scenarios.rst:327]: Yes Sequential, and Disjoint aren't realistic. But they exists, not my fault ;)
Overlap ftw!

File: docs/_tutorials/scenarios/scenarios.rst (hunk 8 above)
Comment (arthurdouillard) [reply docs/_tutorials/scenarios/scenarios.rst:334]: I could write more, but I guess the user is supposed to read a bit about the litterature no?

File: continuum/datasets/segmentation.py (hunk 6 above)
Comment (arthurdouillard) [reply continuum/datasets/segmentation.py]: Done

File: continuum/tasks/task_set.py (hunk 1 above)
Comment (arthurdouillard) [reply continuum/tasks/task_set.py:129]: I'm not really sure about that.

- exterior doesn't need per se to look at the internal, but it's quite straightforward I think. Although we could add in-code doc to explain that.
- currently the additional code to handle different type is quite minimal
- I'm afraid we may confuse more users by having different kind of TaskSets.

File: docs/_tutorials/scenarios/scenarios.rst (hunk 8 above)
Comment (arthurdouillard) [reply docs/_tutorials/scenarios/scenarios.rst:334]: I think the simplest, would be to add a figure to explain the different between the three settings.

File: continuum/tasks/task_set.py (hunk 1 above)
Comment (TLESORT) [reply continuum/tasks/task_set.py:129]: segmentation and classification are two different types of tasks, it makes sens to use different tasksets for both.
A taskset represent a task, different types of taskset for different types of task is quite normal.
I like to have structure in projects and not put to much features or use cases for one class.
I prefer it with two different objects, if you don't have a strong opinion about it and since it is a personal opinion from me I can do it if you want.

File: docs/_tutorials/scenarios/scenarios.rst (hunk 8 above)
Comment (TLESORT) [reply docs/_tutorials/scenarios/scenarios.rst:334]: yes figures would help I think. I don't necessarily think that more should be written but it can maybe be clearer

Comment (arthurdouillard) [reply docs/_tutorials/scenarios/scenarios.rst:334]: #103

File: continuum/tasks/task_set.py (hunk 1 above)
Comment (arthurdouillard) [reply continuum/tasks/task_set.py:129]: #104
